### ISSUE TITLE
Probable issue in demo.ipynb for Video to Text retrieval (Poor result and Random Seed dependent)

### ISSUE THREAD

**Issue by ApoorvFrontera** [OpenGVLab/InternVideo#198]
Hi Team,

First of all, I want to thank the authors for such great work. 

Coming to the demo for Video to Text (V2T) retrieval code provided (https://github.com/OpenGVLab/InternVideo/tree/main/InternVideo2/multi_modality/demo), I had some questions and issues related to it.

1. The paper mentions that for stage 2 fine-tuning to align the video and text, it was mentioned that the text encoder was _initialized_ from BERT-Large implying there is a FT model of BERT that produced more aligned textual embeddings for V2T retrieval. 
```
For the text and speech encoders, we initialize the text encoder and multimodal
decoder using BERT-Large [Devlin et al., 2018]. Specifically, we utilize the initial 19 layers of BERT-Large as the text
encoder, with the subsequent 5 layers equipped with cross-attention layers serving as the multimodal decoder
```
Is this fine-tuned Bert model provided somewhere? As far as I debugged, the BERT large model initialized in the demo was the same as HF default BERT large model. Or if there is an issue in my understanding, please correct me and tell me if I need to change the configs somehow.

2. Another question was that I think the lines are creating randomly initialized linear layers which are destroying the features learned from the vision and text encoders. (https://github.com/OpenGVLab/InternVideo/blob/main/InternVideo2/multi_modality/demo/utils.py#L147C1-L148C68)
```python
self.vision_proj = nn.Linear(self.vision_width, self.embed_dim)
self.text_proj = nn.Linear(self.text_width, self.embed_dim)
```

Does the following line randomize the features learned? This is the reason why changing the random seed gave different results (rand and probability) over different runs.
```python 
_, vfeat = self.encode_vision(frames, test=True)
vfeat = self.vision_proj(vfeat)
```

Waiting for your quick response.
Thanks :)

**Comment by leexinhao** [comment]
1. We use origin Bert weight for init.
2. After training, the `vision proj` wouldn't randomize the features learned, maybe you don't use 'eval' mode so that you get different results over different runs, see https://github.com/OpenGVLab/InternVideo/issues/107#issuecomment-2083240422

**Comment by ApoorvFrontera** [comment]
Thank you so much for the quick reply.
These changes helped a lot. I had just a few short questions:

1. The model in use (InternVideo2-stage2_1b-224p-f4.pt) has "f4" in the suffix. Does that mean it works with four frames extracted from the video? Will changing the  ```num_frames=8``` in the config improve the result since more information is being given or worsen the results since the model is not trained on 8 frames? 

2. Did you also train Stage 2 with 8 or 16 frames and where can I find these models?

3. I read in one of your issues that 1B stage 2 is not pre-trained with the audio encoder. I see [here](https://github.com/OpenGVLab/InternVideo/blob/eca2cdc5a67d7442063d19963515b5bd0feef627/InternVideo2/multi_modality/scripts/evaluation/stage2/zero_shot/1B/config_didemo.py#L25) that you have used audio in the input for zero-shot evaluation of Stage 2 1B model. Can you please explain how does that work and how can I incorporate audio in my usecase with 1B model?

4. For loading and using the CLIP version of the Stage 2 model (OpenGVLab/InternVideo2-CLIP-1B-224p-f8), will it work with the BERT-large or it has been only trained with [InternVL](https://github.com/OpenGVLab/InternVL/tree/main) text encoder? Where do we provide the path for 1B_clip.pth in the configs and where to get the "chinese_alpaca_lora_7b" model? There have been other issues with similar questions pertaining how to use the CLIP model ([Issue 129](https://github.com/OpenGVLab/InternVideo/issues/129)) can you please explain them here?

5. As mentioned in another [Issue 152](https://github.com/OpenGVLab/InternVideo/issues/152), CLIP model is only trained on on CLIP loss and not on matching loss. My task is mainly with Video2Text retrieval, in that case should I stick with the Stage 2 model for better results (assuming I don't want longer text and multilingual capabilities)?
```For InternVideo2-clip-1B, we only use CLIP loss, which is simpler for practical applications.```

Again I want to thank you for the commendable work. Waiting for your quick reply. 
Thanks :)

**Comment by leexinhao** [comment]
> Thank you so much for the quick reply. These changes helped a lot. I had just a few short questions:
> 
> 1. The model in use (InternVideo2-stage2_1b-224p-f4.pt) has "f4" in the suffix. Does that mean it works with four frames extracted from the video? Will changing the  `num_frames=8` in the config improve the result since more information is being given or worsen the results since the model is not trained on 8 frames?
> 2. Did you also train Stage 2 with 8 or 16 frames and where can I find these models?
> 3. I read in one of your issues that 1B stage 2 is not pre-trained with the audio encoder. I see [here](https://github.com/OpenGVLab/InternVideo/blob/eca2cdc5a67d7442063d19963515b5bd0feef627/InternVideo2/multi_modality/scripts/evaluation/stage2/zero_shot/1B/config_didemo.py#L25) that you have used audio in the input for zero-shot evaluation of Stage 2 1B model. Can you please explain how does that work and how can I incorporate audio in my usecase with 1B model?
> 4. For loading and using the CLIP version of the Stage 2 model (OpenGVLab/InternVideo2-CLIP-1B-224p-f8), will it work with the BERT-large or it has been only trained with [InternVL](https://github.com/OpenGVLab/InternVL/tree/main) text encoder? Where do we provide the path for 1B_clip.pth in the configs and where to get the "chinese_alpaca_lora_7b" model? There have been other issues with similar questions pertaining how to use the CLIP model ([Issue 129](https://github.com/OpenGVLab/InternVideo/issues/129)) can you please explain them here?
> 5. As mentioned in another [Issue 152](https://github.com/OpenGVLab/InternVideo/issues/152), CLIP model is only trained on on CLIP loss and not on matching loss. My task is mainly with Video2Text retrieval, in that case should I stick with the Stage 2 model for better results (assuming I don't want longer text and multilingual capabilities)?
>    `For InternVideo2-clip-1B, we only use CLIP loss, which is simpler for practical applications.`
> 
> Again I want to thank you for the commendable work. Waiting for your quick reply. Thanks :)

1. Maybe 8 or 16 is better than 4, but we find that when increase the number of frames to longer (32 or longer), it become worse, this is really a flaw in our current model.
2. No, you can directly use the f4 model for 8 or 16 frames input.
3. In fact, we don't use audio here, it's just a unused config, you could freeze the vit and add a audio encoder, then post-pretraining with audio-text data, video-text data and audio-video text data. We don't do this for InternVideo1B because we focus on vision and audio information is of limited help.
4. See [Readme](https://github.com/OpenGVLab/InternVideo/tree/eca2cdc5a67d7442063d19963515b5bd0feef627/InternVideo2/multi_modality) for chinese_alpaca_lora_7b, which problem is it (to much here);
5. Yes, matching loss benefit performance and s2 model is more lightweight.

**Comment by ApoorvFrontera** [comment]
Thank you so much for all the quick reply and help. This really helped me a lot.
